editors/code/src/commands/inlay_hints.ts

```diff
@@ -21,32 +21,44 @@ const typeHintDecorationType = vscode.window.createTextEditorDecorationType({
 
 export class HintsUpdater {
     private displayHints = true;
+    private drawnDecorations = new Map<string, vscode.DecorationOptions[]>();
```

Comment:
So, we never clear this map unless we turn the hints of? This is (a small) memory leak than. I don't think by itself it is a huge problem, but it hints that the general lifetime cycle is not ideal here. I wonder if we can do better?

I think we can do one of the two things:

* VsCode seems to use subscriptions/disposable pattern to clean up various observers automatically. I wonder if there's a disposable array, associated with the `vscode.TextEditor`, which we can use to react on editor opend/closed.
* alternatively, a pure JS solution would be to use a [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap), keyed by an actual editor here. Not sure if WeakMap is available in VsCode

Reply:
Also, I've seen this problem quite a few times: if you have an observer, that observes an object `X`, it's usually a hard problem to decide when you should stop observing (because there's a reference cycle between observer and an observee). I wonder if there's a blog post about this somewhere...

Reply:
Nice catch, had forgotten to tell my concerns about it.

There's a `onDidCloseTextDocument` that is triggered whenever a tab is closed, but that does not help in the current situation much: the whole idea is to quickly redraw the decorations when it's reopened, so we cannot remove them from the `drawnDecorations` yet.

`WeakMap` sounds like a nice idea, I will try it out.

Reply:
Looks like both `Map` and `WeakMap` compare `vscode.Uri` instances by reference and those change every time we close and open the editor tab for the file.
The uri string stays the same, but `WeakMap` does not accept `string` types as keys, so I had to use the `Map` again.

But now, since we clear the map every time an edit is made, the leak gets even smaller, doesn't it?